prow/config/inrepoconfig.go

```diff
@@ -187,3 +188,46 @@ func DefaultAndValidateProwYAML(c *Config, p *ProwYAML, identifier string) error
 
 	return utilerrors.NewAggregate(errs)
 }
+
+type InRepoConfigGitCache struct {
+	git.ClientFactory
+	cache map[string]git.RepoClient
+	lock  sync.Mutex
+}
+
+func NewInRepoConfigGitCache(factory git.ClientFactory) git.ClientFactory {
+	return &InRepoConfigGitCache{
+		ClientFactory: factory,
+		cache:         map[string]git.RepoClient{},
+	}
+}
+
+func (c *InRepoConfigGitCache) ClientFor(org, repo string) (git.RepoClient, error) {
```

Comment:
I am not really familiar with `sub`, but this will only work if it never modifies the clone. It is also not quite obvious how this is specific to `InRepoConfig`?

Lets please either move this into sub or have a big fat go doc warning that the clients must not be used to modify the clone and that the files could be modified while reading them (if someone else calls `ClientFor`)

Reply:
> this will only work if it never modifies the clone

You mean editing files, right? I don't think checking out and merging various refs as `inrepoconfig` does should be a problem.

This may not be have to be specific to `InRepoConfig`, that is just the use case I have in mind and I chose to name it specifically because I am optimizing for that specific use case. If this ends up being generally useful I'm happy to choose a more generic name.

I'm realizing that what I have written doesn't provide sufficient synchronization though. I'm going to hold this and iterate. The next version may be more obviously coupled to `inrepo` config.
/hold

Reply:
Yeah I meant modifying files/leaving an unclean working directory.

Inrepoconfig doesn't do that but the resulting client is still not threadsafe, as the checkouts manipulate the same dir.

Reply:
Yeah the threadsafety issue was what I mean by insufficient synchronization :slightly_smiling_face: 

I think I've resolved that issue and I've added a test to validate that the concurrent behavior is correct. PTAL.
/hold cancel